scp/gen_operators/SCPOperatorSysGen.cpp

```diff
@@ -0,0 +1,270 @@
+
+    if (!(operand->GetOrder() > 0 && operand->GetOrder() < 5 && operands[operand->GetOrder() - 1] == nullptr))
+    {
+#ifdef SCP_DEBUG
+      std::stringstream ss;
+      ss << "Invalid operand order ";
+      ss << (int)operand->GetOrder();
+      Utils::logSCPError(ms_context, ss.str(), addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+    else
+    {
+      operands[operand->GetOrder() - 1] = operand;
+    }
+  }
+
+  for (size_t i = 0; i < 3; ++i)
+  {
+    auto * operand = operands[i];
+    if (!operand)
+    {
+#ifdef SCP_DEBUG
+      Utils::logSCPError(ms_context, "One or more operands missed", addr);
+      Utils::printInfo(ms_context, addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+  }
+  return SC_RESULT_OK;
+}
+
+sc_result SCPOperatorSysGen::Execute()
+{
+  if (operands[0]->IsAssign())
+  {
+#ifdef SCP_DEBUG
+    Utils::logSCPError(ms_context, "Operand 1 must have FIXED modifier", addr);
+#endif
+    FinishExecutionWithError();
+    return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
+
+  if (operands[2]->IsAssign())
+  {
+#ifdef SCP_DEBUG
+    Utils::logSCPError(ms_context, "Operand 3 must have FIXED modifier", addr);
+#endif
+    FinishExecutionWithError();
+    return SC_RESULT_ERROR_INVALID_PARAMS;
+  }
+
+  if (operands[0]->IsFixed())
+  {
+    if (!operands[0]->GetValue().IsValid())
+    {
+#ifdef SCP_DEBUG
+      Utils::logSCPError(ms_context, "Operand 1 has modifier FIXED, but has no value", addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+  }
+
+  if (operands[2]->IsFixed())
+  {
+    if (!operands[2]->GetValue().IsValid())
+    {
+#ifdef SCP_DEBUG
+      Utils::logSCPError(ms_context, "Operand 3 has modifier FIXED, but has no value", addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+  }
+
+  ScTemplateParams generationParams;
+  ScAddr const & generationParamsAddr = operands[2]->GetValue();
+  ScIterator3Ptr const paramsMapIt3 = ms_context->Iterator3(
+      generationParamsAddr, ScType::EdgeAccessConstPosPerm, ScType::NodeConst);
+
+  while (paramsMapIt3->Next())
+  {
+    ScAddr const & parameterPairAddr = paramsMapIt3->Get(2);
+
+    ScIterator5Ptr const varIt5 = ms_context->Iterator5(
+        parameterPairAddr, ScType::EdgeAccessConstPosPerm, ScType::Unknown, ScType::EdgeAccessConstPosPerm, Keynodes::rrel_1);
+
+    ScAddr varAddr;
+    if (varIt5->Next())
+      varAddr = varIt5->Get(2);
+    else
+    {
+#ifdef SCP_DEBUG
+      Utils::logSCPError(ms_context, "Generation parameters pair must have variable", addr);
+#endif
+      FinishExecutionWithError();
+      return SC_RESULT_ERROR_INVALID_PARAMS;
+    }
+
+    ScIterator5Ptr const replacementIt5 = ms_context->Iterator5(
+        parameterPairAddr, ScType::EdgeAccessConstPosPerm, ScType::Unknown, ScType::EdgeAccessConstPosPerm, Keynodes::rrel_2);
```

Comment:
why not basic sequence?

Reply:
Backward compatibility with old scp-machine